### PR TITLE
fix: snippet suggestion does not replace the input string

### LINKED ISSUE #5926: snippet suggestion does not replace the input string
### Description
snippet suggestion does not replace the input string, just append snippet to it

### Reproduction Steps
autocompletion1

**OS and Theia version:**
ubuntu 18.04
theia 0.9.0
chrome 76.0.3809.100

**Diagnostics:**

### PR DESCRIPTION
**Pull request by yeweiasia**
#### What it does

fix https://github.com/theia-ide/theia/issues/5926: snippet suggestion does not replace the input string

- [x] a CQ for copied code: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20668 (please mark when it is resolved or labeled with `checkin` ready

#### How to test

1.theia 0.9.0 with [redhat.java-0.46.0.vsix](https://github.com/redhat-developer/vscode-java/releases/download/v0.46.0/redhat.java-0.46.0.vsix)
2.open a maven project and edit a java file
3.type keyword like "if".
image
4.select "ifelse" or other snippet suggestion.
5.without this patch whole suggested item body will be append to "if"  which is demonstrated in https://github.com/theia-ide/theia/issues/5926
autocompletion1
6.with this patch "if" will  replaced by suggested item body.
autocompletion2
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by yeweiasia** [comment]
https://github.com/theia-ide/theia/issues/5926

**Comment by vince-fugnitto** [comment]
@yeweiasia can you provide comment on the PR template, mainly 'what it does' and 'how to test'?

**Comment by yeweiasia** [comment]
@vince-fugnitto comment updated

File: packages/monaco/src/browser/monaco-snippet-suggest-provider.ts [hunk 1]
```diff
@@ -35,6 +35,8 @@ export class MonacoSnippetSuggestProvider implements monaco.modes.ISuggestSuppor
     }
 
     resolveCompletionItem(_: monaco.editor.ITextModel, __: monaco.Position, item: monaco.modes.ISuggestion): monaco.modes.ISuggestion {
+        const word = _.getWordAtPosition(__);
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: please rename to proper names, underscore means that parameters are not used

used variables should follow naming guidelines: https://github.com/theia-ide/theia/wiki/Coding-Guidelines#whole-words-names

**Comment by akosyakov** [comment]
You can have a look how VS Code does it here: https://github.com/TypeFox/vscode/blob/monaco/0.18.0/src/vs/workbench/contrib/snippets/browser/snippetCompletionProvider.ts#L80-L140

**Comment by yeweiasia** [comment]
ok, I will investigate how vscode implemented in next few days.

**Comment by yeweiasia** [comment]
autocompletion3
@akosyakov PR updated.

File: packages/monaco/src/browser/monaco-snippet-suggest-provider.ts [hunk 2]
```diff
@@ -24,17 +24,66 @@ export class MonacoSnippetSuggestProvider implements monaco.modes.ISuggestSuppor
...
     protected readonly pendingSnippets = new Map<string, Promise<void>[]>();
 
-    async provideCompletionItems(model: monaco.editor.ITextModel): Promise<monaco.modes.ISuggestResult> {
+    async provideCompletionItems(model: monaco.editor.ITextModel, position: monaco.Position, context: monaco.modes.SuggestContext): Promise<monaco.modes.ISuggestResult> {
         const languageId = model.getModeId(); // TODO: look up a language id at the position
         await this.loadSnippets(languageId);
-        const suggestions = this.snippets.get(languageId) || [];
+        const snippetsForLanguage = this.snippets.get(languageId) || [];
+
+        // copied and modified from https://github.com/microsoft/vscode/blob/master/src/vs/workbench/contrib/snippets/browser/snippetCompletionProvider.ts
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: We need a CQ for it. I will help with it. Unfortunately, CQs resolution can take sometime and we cannot merge before :(

Comment (akosyakov) [reply packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: We need to add MS copyright under ours copyright as well.

**Comment by akosyakov** [comment]
@yeweiasia Does it address https://github.com/theia-ide/theia/issues/4208 as well? Could you update the description with `fix` if so.

**Comment by akosyakov** [comment]
I've updated a description with a link to the CQ.

**Comment by yeweiasia** [comment]
@akosyakov I think this can be implemented in my own way, I will rewrite and make another PR to resolve this bug

**Comment by yeweiasia** [comment]
> @yeweiasia Does it address #4208 as well? Could you update the description with `fix` if so.

I will test after rewriting

**Comment by akosyakov** [comment]
> @akosyakov I think this can be implemented in my own way, I will rewrite and make another PR to resolve this bug

What for? It will make it harder to keep in sync for future updates. We can wait for a CQ.

**Comment by yeweiasia** [comment]
@akosyakov I have implemented another version of this feature, https://github.com/theia-ide/theia/pull/5963
this version is more concise when dealing with pattern, I just find out the word at the position rather than find out all offsets of the word in whole line. I think find out all word in whole line make no sense in case of snippet code suggestion.

**Comment by akosyakov** [comment]
@yeweiasia if you are sure that your implementation is better than let's go with it. We can reconsider later to use VS Code implementation if it will be required. Please close this PR then.

File: packages/monaco/src/browser/monaco-snippet-suggest-provider.ts (hunk 2 above)
Comment (yeweiasia) [reply packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: ms copyright added

**Comment by yeweiasia** [comment]
@akosyakov after making a try to provide another implementation, It seems better to sync with vscode logic and I have fully synced with vscode impl :)

**Comment by akosyakov** [comment]
@BroKun @vince-fugnitto Could you finish the review on this PR please?

**Comment by akosyakov** [comment]
CQ was marked as ready for `checkin`, @BroKun @vince-fugnitto please merge if you are happy after testing

**Pull request merged**